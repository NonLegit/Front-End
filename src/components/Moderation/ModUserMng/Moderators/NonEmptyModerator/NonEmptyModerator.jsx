/* eslint-disable react/jsx-no-constructed-context-values */
/* eslint-disable import/no-cycle */
import {
  Box, Typography, IconButton,
} from '@mui/material';
import EditIcon from '@mui/icons-material/Edit';
import * as React from 'react';
import DeleteIcon from '@mui/icons-material/Delete';
import { useNavigate } from 'react-router-dom';
import { StyledAvatar, UserBar, UserContainer } from '../../styles';
import RemovePopUp from '../RemovePopUp/RemovePopUp';
import EditPopUp from '../EditPopUp/EditPopUp';

export const RemoveContext = React.createContext();
export const EditContext = React.createContext();

/**
 * non empty moderator page
 * @component
 * @property {string} username the username of the moderator
 * @property {string} profilePicture it is the link to the profile picture of the moderator
 * @property {string} modDate it is the date to be a moderator
 * @property {boolean} all indicates that the moderator has all privileges
 * @property {boolean} access indicates that tha moderator has the control on user
 * @property {boolean} confing indicates that the moderator has the control on settings
 * @property {boolean} flair indicates that the moderator has the control on post flairs
 * @property {boolean} posts indicates that the moderator has the control on posts
 * @property {integer} type determine the type of moderators if invited or arleady a moderator
 * @return {React.Component} - non empty moderator page component
 */

function NonEmptyModerator(props) {
  // 0 stands for all moderators (marked as editable but handeled in database)
  // 1 stands for Invited moderators
  const {
    userName, profilePicture, modDate, all, access, config, flair, posts, type,
  } = props;

  const navigate = useNavigate();

  // navigate
  const handleClickUser = (userName) => {
    navigate(`/user/${userName}`);
  };

  const [openRemove, setOpenRemove] = React.useState(false);
  const handleClickOpenRemove = () => { setOpenRemove(true); };
  const handleClickCloseRemove = () => { setOpenRemove(false); };

  const [openEdit, setOpenEdit] = React.useState(false);
  const handleClickOpenEdit = () => { setOpenEdit(true); };
  const handleClickCloseEdit = () => { setOpenEdit(false); };

  const handlePermissions = () => {
    const permissions = [];
    if (all === true) permissions.push('Everything');
    else {
      if (access === true) permissions.push('Manage Users, ');
      if (config === true) permissions.push('Manage Settings, ');
      if (flair === true) permissions.push('Manage Flair, ');
      if (posts === true) permissions.push('Manage Posts & Comments, ');
      const lastElemnt = permissions.pop().trimEnd();
      permissions.push(lastElemnt.substring(0, lastElemnt.length - 1));
    }
    return permissions;
  };

  return (
    <UserBar>
      <RemoveContext.Provider value={{
        openRemove, handleClickCloseRemove, userName,
      }}
      >
        <RemovePopUp />
      </RemoveContext.Provider>
      <EditContext.Provider value={{
        openEdit, handleClickCloseEdit,
      }}
      >
        <EditPopUp userName={userName} />
      </EditContext.Provider>
      <UserContainer onClick={() => handleClickUser(userName)}>
        <StyledAvatar src={profilePicture} variant="square" />
        <Typography
          padding="8px"
          fontSize="15px"
          fontWeight="bold"
        >
          {userName}
        </Typography>
      </UserContainer>
      <Box display="flex">
        <Typography
          padding="8px"
          fontSize="12px"
          color="#878A8C"
          width="max-content"
        >
          {modDate}
        </Typography>
      </Box>
      <Box sx={{
        alignItems: 'center', display: 'flex', padding: '8px 16px', width: '100%', justifyContent: 'flex-end',
      }}
      >
        {(() => {
          switch (type) {
            case '1': return (
              <>
                <Typography
                  padding="4px"
                  fontSize="12px"
                  color="#878A8C"
                >
                  {handlePermissions()}
                </Typography>
                <IconButton disableRipple disableTouchRipple onClick={handleClickOpenRemove}>
                  <DeleteIcon fontSize="small" />
                </IconButton>
              </>
            );
            default: return (
              <>
                <Typography
                  padding="4px"
                  fontSize="12px"
                  color="#878A8C"
                >
                  {handlePermissions()}
                </Typography>
                <IconButton disableRipple disableTouchRipple onClick={handleClickOpenEdit}>
                  <EditIcon fontSize="small" />
                </IconButton>
              </>
            );
          }
        })()}
      </Box>
    </UserBar>
  );
}
export default NonEmptyModerator;
