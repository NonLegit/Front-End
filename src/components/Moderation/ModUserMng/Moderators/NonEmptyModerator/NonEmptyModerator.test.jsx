import renderer from 'react-test-renderer';
import NonEmptyModerator from './NonEmptyModerator';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

test('test snapshot', async () => {
  const tree = renderer.create(
    <NonEmptyModerator
      userName="fady"
      profilePicture="users/default.png"
      modDate="2022-12-20T00:53:55.071Z"
      all
      access
      config
      flair
      posts
      type={1}
    />,
  )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
