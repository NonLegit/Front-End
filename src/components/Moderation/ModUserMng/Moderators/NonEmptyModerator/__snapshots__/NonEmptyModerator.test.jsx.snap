// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<div
  className="MuiBox-root css-13en8op"
>
  <div
    className="MuiBox-root css-1gwqxiu"
    onClick={[Function]}
  >
    <div
      className="MuiAvatar-root MuiAvatar-square css-1v1kfpo-MuiAvatar-root"
    >
      <img
        className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
        src="users/default.png"
      />
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-gu63a4-MuiTypography-root"
    >
      fady
    </p>
  </div>
  <div
    className="MuiBox-root css-k008qs"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-177droq-MuiTypography-root"
    >
      2022-12-20T00:53:55.071Z
    </p>
  </div>
  <div
    className="MuiBox-root css-1uygd4d"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1keaqvn-MuiTypography-root"
    >
      Everything
    </p>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-yvetkp-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="EditIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </button>
  </div>
</div>
`;
